From 2ba2eef5b11d81b801265ae92df91571c19cc8a3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vincent=20Stehl=C3=A9?= <vincent.stehle@arm.com>
Date: Tue, 30 Sep 2025 11:39:54 +0200
Subject: [PATCH] Update linker flags
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The -I$(XEN_ROOT)/tools/include added to QEMU's extra-cflags causes
__XEN_INTERFACE_VERSION__ to be defined before xen_native.h is included,
triggering an include-order assertion. Downgrade to a warning since the
version is consistent in cross-compile.
Ref: https://github.com/qemu/qemu/commit/e2abfe5ec6

Upstream: Buildroot specific. This is a temporary patch for an intermediate bump.
Signed-off-by: Vincent Stehlé <vincent.stehle@arm.com>
---
 tools/Makefile                              | 29 ++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletions(-)

diff --git a/tools/Makefile b/tools/Makefile
index a9a3196e..5d289651 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -37,8 +37,12 @@ SUBDIRS-$(CONFIG_X86) += xenpaging
 SUBDIRS-$(CONFIG_X86) += debugger
 SUBDIRS-$(CONFIG_TESTS) += tests
 
+# These don't cross-compile
+ifeq ($(XEN_COMPILE_ARCH),$(XEN_TARGET_ARCH))
 SUBDIRS-y += python
 SUBDIRS-$(CONFIG_PYGRUB) += pygrub
+endif
+
 SUBDIRS-$(OCAML_TOOLS) += ocaml
 
 ifeq ($(CONFIG_RUMP),y)
@@ -189,8 +193,31 @@ subdir-all-qemu-xen-dir: qemu-xen-dir-find
 		-DXC_WANT_COMPAT_GNTTAB_API=1 \
 		-DXC_WANT_COMPAT_MAP_FOREIGN_API=1 \
 		-DXC_WANT_COMPAT_DEVICEMODEL_API=1 \
+		-I$(XEN_ROOT)/tools/include \
 		$(EXTRA_CFLAGS_QEMU_XEN)" \
-		--extra-ldflags="$(QEMU_UPSTREAM_RPATH)" \
+		--extra-ldflags=" \
+		-L$(XEN_ROOT)/tools/libs/call \
+		-L$(XEN_ROOT)/tools/libs/ctrl \
+		-L$(XEN_ROOT)/tools/libs/devicemodel \
+		-L$(XEN_ROOT)/tools/libs/evtchn \
+		-L$(XEN_ROOT)/tools/libs/foreignmemory \
+		-L$(XEN_ROOT)/tools/libs/gnttab \
+		-L$(XEN_ROOT)/tools/libs/guest \
+		-L$(XEN_ROOT)/tools/libs/store \
+		-L$(XEN_ROOT)/tools/libs/toolcore \
+		-L$(XEN_ROOT)/tools/libs/toollog \
+		-L$(XEN_ROOT)/tools/xenstore \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/call \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/ctrl \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/devicemodel \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/evtchn \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/foreignmemory \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/gnttab \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/guest \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/store \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/toolcore \
+		-Wl,-rpath-link=$(XEN_ROOT)/tools/libs/toollog \
+		$(QEMU_UPSTREAM_RPATH)" \
 		--bindir=$(LIBEXEC_BIN) \
 		--datadir=$(SHAREDIR)/qemu-xen \
 		--localstatedir=$(localstatedir) \
-- 
2.51.0

